src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/SortRewriter.cs

```diff
@@ -42,14 +42,50 @@ public override Expression VisitSqlRoot(SqlRootExpression expression, SearchOpti
                 return expression;
             }
 
+            bool matchFound = false;
+            for (int i = 0; i < expression.SearchParamTableExpressions.Count; i++)
+            {
+                SearchParamTableExpression currentExpression = expression.SearchParamTableExpressions[i];
+                Expression updatedExpression = currentExpression.Predicate.AcceptVisitor(this, context);
+                if (ReferenceEquals(currentExpression.Predicate, updatedExpression))
+                {
+                    continue;
+                }
+                else
+                {
+                    matchFound = true;
+                    break;
+                }
+            }
+
+            SearchParamTableExpressionKind sortKind = matchFound ? SearchParamTableExpressionKind.SortWithFilter : SearchParamTableExpressionKind.Sort;
+
             var queryGenerator = _searchParamTableExpressionQueryGeneratorFactory.GetSearchParamTableExpressionQueryGenerator(context.Sort[0].searchParameterInfo);
 
             var newTableExpressions = new List<SearchParamTableExpression>(expression.SearchParamTableExpressions.Count + 1);
             newTableExpressions.AddRange(expression.SearchParamTableExpressions);
 
-            newTableExpressions.Add(new SearchParamTableExpression(queryGenerator, new SortExpression(context.Sort[0].searchParameterInfo), SearchParamTableExpressionKind.Sort));
+            newTableExpressions.Add(new SearchParamTableExpression(queryGenerator, new SortExpression(context.Sort[0].searchParameterInfo), sortKind));
 
             return new SqlRootExpression(newTableExpressions, expression.ResourceTableExpressions);
         }
+
+        public override Expression VisitSearchParameter(SearchParameterExpression expression, SearchOptions context)
+        {
+            if (context.Sort.Count > 0)
+            {
+                if (expression.Parameter.Equals(context.Sort[0].searchParameterInfo))
+                {
+                    return new SearchParameterExpression(expression.Parameter, expression.Expression);
```

Comment:
This allocation is not used anywhere?

Reply:
My idea was to use this just as a "signal" to the caller that we have a `SearchParameterExpression` with a search parameter that is the same as the sort parameter. 

---
In reply to: [613560818](https://github.com/microsoft/fhir-server/pull/1797#discussion_r613560818) [](ancestors = 613560818)

Reply:
Right. You could use a sentinel value for that

Reply:
Updated code to return null and avoid the extra allocation.